common/src/main/java/io/netty/util/concurrent/LimitLatch.java

```diff
@@ -0,0 +1,111 @@
+package io.netty.util.concurrent;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.AbstractQueuedSynchronizer;
+
+/**
+ * Shared latch that allows the latch to be acquired a limited number of times
+ * after which all subsequent requests to acquire the latch will fail.
+ */
+public class LimitLatch {
+
+    private class Sync extends AbstractQueuedSynchronizer {
+        private static final long serialVersionUID = 1L;
+
+        public Sync() {
+        }
+
+        @Override
+        protected int tryAcquireShared(int ignored) {
+            long newCount = count.incrementAndGet();
+            if (newCount > limit) {
+                // Limit exceeded
+                count.decrementAndGet();
+                return -1;
+            } else {
+                return 1;
+            }
+        }
+
+        @Override
+        protected boolean tryReleaseShared(int arg) {
+            count.decrementAndGet();
+            return true;
+        }
+    }
+
+    private final Sync sync;
+    private final AtomicLong count;
+    private volatile long limit;
+
+    /**
+     * Instantiates a LimitLatch object with an initial limit.
+     *
+     * @param limit - maximum number of concurrent acquisitions of this latch
+     */
+    public LimitLatch(long limit) {
+        this.limit = limit;
+        this.count = new AtomicLong(0);
+        this.sync = new Sync();
+    }
+
+    /**
+     * Returns the current count for the latch
+     *
+     * @return the current count for latch
+     */
+    public long getCount() {
+        return count.get();
+    }
+
+    /**
+     * Obtain the current limit.
+     */
+    public long getLimit() {
+        return limit;
+    }
+
+
+    /**
+     * Sets a new limit. If the limit is decreased there may be a period where
+     * more shares of the latch are acquired than the limit. In this case no
+     * more shares of the latch will be issued until sufficient shares have been
+     * returned to reduce the number of acquired shares of the latch to below
+     * the new limit. If the limit is increased, threads currently in the queue
+     * may not be issued one of the newly available shares until the next
+     * request is made for a latch.
+     *
+     * @param limit The new limit
+     */
+    public void setLimit(long limit) {
+        this.limit = limit;
+    }
+
+
+    /**
+     * Acquires a shared latch if one is available or waits for one if no shared
+     * latch is current available.
+     */
+    public void countUpOrAwait() throws InterruptedException {
+        sync.acquireSharedInterruptibly(1);
+    }

```

Comment:
it truly is dangerous to block the loop. the reason i add keep this method is 
1. because there are incoming in the backlog queue, select() would block, so the loop would keep spinning. this blocking method is meant to slove this "busy loop" problem.
2. i made the assumption that we won't run task in boss loop.
3. another way to solve "busy loop" problem would be make io.netty.channel.nio.AbstractNioChannel.NioUnsafe#read() return a special value to indicate that we have read too much, so the loop may choose to pause a little while if there is no task to run. java.lang.Integer#MAX_VALUE would be perfect for this.